### PR TITLE
Adds lag compensation for the supermatter zaps and some antagonists.

### PR DESCRIPTION
**Pull request by Pickle-Coding**
## About The Pull Request
Adds delta time scaling for changeling chemical generation, xenomorph plasma generation, xenomorph resin healing (up to 8 seconds per tick to prevent weird instant heal phenomena in extreme cases), and revenant essence generation. This delta time scaling is based on the world's time and the respective subsystem's last fire. This will help keep these generation smooth even when their respective subsystem lags, preventing lag from ruining the antagonists rounds. This can still slow down if the actual world is slowing down, but that's fine because the rest of the world will typically be slowing down proportionately so it's fair.

Fixes supermatter zap rate not scaling with its internal energy properly. Ghilker forgot to scale the time reduction by seconds. It should be much faster at higher energy levels, making the transition between 0 and 5 GeV more smooth.

Supermatter zap power generation will scale by the world's delta time between the supermatter's last zap, making power generation more consistent even when atmos lags. Also prevents the supermatter zap rate scaling from also scaling the power generation by extreme amounts, keeping it linear and consistent for the first 5 GeV (>5GeV has cringe power multipliers, but I'm going to ignore that for this PR, because I want this to mainly be a consistency thing rather than a balance thing). A 1.5GeV supermatter will be able to generate around 800kW* of power. I forgot exactly what goal of power we want the roundstart SM to produce, so I might change the scale if I find the target power generation.

* - By generate 800kW of power I mean the energy the actual zaps contain. Tesla coils don't collect 100% of it. I'll test later to see how much an actual roundstart SM will output to grid (ignoring SMES), and I'll try to keep it as close as possible to what it was before this PR assuming no lag.
## Why It's Good For The Game
Subsystems can disproportionately lag a lot more compared to the rest of the subsystems and game. This can ruin antagonist rounds as it can slow their chemicals, plasma or whatever generation grinding to a halt. The delta time scaling will significantly reduce the impact of lag for these antagonists, allowing them to play more normally even while the subsystems their generators run on are lagging.

Supermatter zap rate was supposed to fire every 2.5 or something seconds at 1.5GeV according to comments on older commits. I just assume they accidentally forgot to scale their time reduction by seconds, as doing that made it work accordingly.

For supermatter zap power generation delta time scaling, kinda same reasoning as for the antagonists. We don't want power generation to stop because atmos ss is being slow. Scaling between the zaps also prevents a nonlinear boost to power generation making it even more absurd at higher energy levels.
## Changelog
:cl:
balance: Supermatter zap power generation scales with the delta time between its last zaps, preventing faster zapping from scaling power generation to extreme levels.
fix: Fixes supermatter zap rate not scaling properly. It should zap much faster at higher energy levels as intended.
qol: Changeling chemical generation scales with the world's delta time, making its rate independent of subsystem lag.
qol: Revenant essence generation scales with the world's delta time, making its rate independent of subsystem lag.
qol: Xenomorph plasma generation and resin healing scales with the world's delta time, making their rates independent of subsystem lag.
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
hey bro your PR title should be short.

**Comment by Sealed101** [comment]
our daddy taught us not be ashamed of our PR titles, especially since they're such good size and all

File: code/__DEFINES/subsystems.dm [hunk 1]
```diff
@@ -321,6 +321,9 @@
 #define SSMOBS_DT (SSmobs.wait/10)
 #define SSOBJ_DT (SSobj.wait/10)
 
+// The change in the world's time from the subsystem's last fire in seconds.
+#define DELTA_WORLD_TIME(ss) ((world.time - ss.last_fire) / 10)
```
Comment (Pickle-Coding) [inline code/__DEFINES/subsystems.dm:325]:
```suggestion
#define DELTA_WORLD_TIME(ss) ((world.time - ss.last_fire) * 0.1)
```

File: code/modules/power/supermatter/supermatter.dm [hunk 2]
```diff
@@ -271,16 +271,19 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
...
 	internal_energy_factors = calculate_internal_energy()
 	zap_factors = calculate_zap_multiplier()
-	if(internal_energy && (last_power_zap + 4 SECONDS - (internal_energy * 0.001)) < world.time)
+	if(internal_energy && (last_power_zap + (4 - internal_energy * 0.001) SECONDS) < world.time)
 		playsound(src, 'sound/weapons/emitter2.ogg', 70, TRUE)
 		hue_angle_shift = clamp(903 * log(10, (internal_energy + 8000)) - 3590, -50, 240)
 		var/zap_color = color_matrix_rotate_hue(hue_angle_shift)
+		//Scale the strength of the zap with the world's time elapsed between zaps in seconds.
+		//Capped at 16 seconds to prevent a crazy burst of energy if atmos was halted for a long time.
+		var/delta_time = min((world.time - last_power_zap) / 10, 16)
```
Comment (Pickle-Coding) [inline code/modules/power/supermatter/supermatter.dm:280]:
```suggestion
		var/delta_time = min((world.time - last_power_zap) * 0.1, 16)
```

**Review by Ghommie — APPROVED** [review]

**Pull request merged**